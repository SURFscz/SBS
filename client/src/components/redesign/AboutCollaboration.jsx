import React from "react";
import I18n from "i18n-js";
import "./AboutCollaboration.scss";
import {ReactComponent as ServicesIcon} from "../../icons/services.svg";
import {ReactComponent as TerminalIcon} from "../../icons/terminal.svg";
import {ReactComponent as IllustrationCO} from "../../icons/illustration-CO.svg";
import {removeDuplicates, stopEvent} from "../../utils/Utils";
import Logo from "./Logo";
import {isUserAllowed, ROLES} from "../../utils/UserRole";
import {Tooltip} from "@surfnet/sds";
import {CO_SHORT_NAME, SRAM_USERNAME} from "../../validations/regExps";
import {clearFlash} from "../../utils/Flash";

const memberCutOff = 10;

class AboutCollaboration extends React.Component {

    constructor(props, context) {
        super(props, context);
        this.state = {
            showAll: false
        };
    }

    openMembersDetails = e => {
        stopEvent(e);
        const {tabChanged} = this.props;
        tabChanged("members");
    }

    openService = service => e => {
        stopEvent(e);
        clearFlash();
        this.props.history.push(`/services/${service.id}`);
    }

    formatServiceUri = (serviceUri, collaboration, user) => {
        if (serviceUri.indexOf(CO_SHORT_NAME) > -1) {
            serviceUri = serviceUri.replace(CO_SHORT_NAME, collaboration.short_name);
        }
        if (serviceUri.indexOf(SRAM_USERNAME) > -1) {
            serviceUri = serviceUri.replace(SRAM_USERNAME, user.username);
        }
        return serviceUri;
    }

    openServiceUri = (service, collaboration, user) => () => {
        service.uri && window.open(this.formatServiceUri(service.uri, collaboration, user), "_blank");
    }

    toggleShowMore = e => {
        stopEvent(e);
        this.setState({showAll: !this.state.showAll})
    }

    render() {
        const {showAll} = this.state;
        const {collaboration, user, showMemberView, isJoinRequest} = this.props;
        const isAllowedToSeeMembers = !isJoinRequest &&
            isUserAllowed(ROLES.COLL_ADMIN, user, collaboration.organisation_id, collaboration.id) && !showMemberView;
        const services = isJoinRequest ? [] : removeDuplicates(collaboration.services.concat(collaboration.organisation.services), "id");
        const {collaboration_memberships} = collaboration
        const memberships = isJoinRequest ? [] : collaboration_memberships
            .sort((m1, m2) => m1.role.localeCompare(m2.role))
            .slice(0, showAll ? collaboration_memberships.length : memberCutOff);
        const showMembers = !isJoinRequest && (collaboration.disclose_member_information || isAllowedToSeeMembers);
        return (
            <div className="collaboration-about-mod">
                <div className="about">
                    <h2>{I18n.t("models.collaboration.about")}</h2>
                    <p>{collaboration.description}</p>
                </div>
                <div className="details">
                    {services.length > 0 && <div className="services">
                        <h2>{I18n.t("models.collaboration.services", {nbr: services.length})}</h2>
                        {!isJoinRequest && <span className="info">
                        {I18n.t("models.collaboration.servicesStart")}
                        </span>}
                        <ul className="services">
                            {services.sort((a, b) => a.name.localeCompare(b.name)).map(service =>
                                <div className="service-button" key={service.name}>
                                    <li onClick={this.openServiceUri(service, collaboration, user)}
                                        className={`${service.uri ? "uri" : ""}`}>
                                        {service.logo && <Logo src={service.logo} alt={service.name}/>}
                                        <span className="border-left">{service.name}</span>
                                        {service.uri &&
                                        <span className="border-left no-border open-service">
                                        <Tooltip children={service.uri.startsWith("http") ? <ServicesIcon/> :
                                                     <TerminalIcon/>}
                                                 standalone={true}
                                                 tip={I18n.t("models.collaboration.servicesHoover",
                                                     {uri: this.formatServiceUri(service.uri, collaboration, user)})}/>
                                        </span>}
                                    </li>
                                    <div className="service-links">
                                        {!isJoinRequest &&
                                        <a href={`/services/${service.id}`}
                                           onClick={this.openService(service)}>{I18n.t("models.collaboration.instructions")}</a>}
                                        {service.accepted_user_policy &&
                                        <a href={service.accepted_user_policy} rel="noopener noreferrer"
                                           target="_blank">{I18n.t("service.accepted_user_policy")}</a>
                                        }
                                        {service.privacy_policy &&
                                        <a href={service.privacy_policy} rel="noopener noreferrer"
                                           target="_blank">{I18n.t("service.privacy_policy")}</a>
                                        }
                                    </div>
                                </div>
                            )}
                        </ul>
                    </div>}
                    {services.length === 0 &&
                    <div className="services">
                        {(services.length === 0 && !isJoinRequest) &&
                        <h2>{I18n.t("models.collaboration.noServices")}</h2>}
                        {isJoinRequest && <h2>{I18n.t("models.collaboration.noServicesJoinRequest")}</h2>}
                    </div>}
                    {showMembers &&
                    <div className="members">
                        <div className="members-header">
                            <h2>{I18n.t("models.collaboration.members", {nbr: collaboration.collaboration_memberships.length})}</h2>
                            <a href="/details"
                               onClick={this.openMembersDetails}>{I18n.t("models.collaboration.showMemberDetails")}</a>
                        </div>
                        <ul>
                            {memberships.map(m => <li key={m.id}>
                                <span className="member">
                                    {m.user.name}
                                    {m.role === "admin" &&
                                    <span className="role">{` (${I18n.t("models.collaboration.admin")})`}</span>}
                                </span>
                            </li>)}
                        </ul>
                        {collaboration.collaboration_memberships.length > memberCutOff &&
                        <a href={showAll ? "/more" : "/less"}
                           onClick={this.toggleShowMore}>{I18n.t(`models.collaboration.${showAll ? "less" : "more"}`,
                            {nbr: collaboration_memberships.length - memberCutOff})}</a>}

                    </div>}
                    {showMembers && <div className="playing-svg">
                        <IllustrationCO/>
                    </div>}
                    {(!showMembers && !isJoinRequest) && <div className="members">
                        <div className="members-header">
                            <h2>{I18n.t("models.collaboration.discloseNoMemberInformation")}</h2>
                        </div>
                    </div>}
                    {isJoinRequest && <div className="members">
                        <div className="members-header join-request">
                            <h2>{I18n.t("models.collaboration.members", {nbr: collaboration.collaboration_memberships.length})}</h2>
                            <p>{I18n.t("models.collaboration.discloseNoMemberInformationJoinRequest")}</p>
                        </div>
                    </div>}
                </div>
            </div>
        );
    }

}

export default AboutCollaboration;