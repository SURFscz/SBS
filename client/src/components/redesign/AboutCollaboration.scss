@import "../../stylesheets/vars";

div.collaboration-about-mod {
    padding: 32px 0;
    max-width: 1280px;
    width: 100%;
    margin: 0 auto;

    div.about {
        margin-bottom: 40px;
    }

    h1 {
        font-size: 22px;
        line-height: 26px;
        margin-bottom: 10px;
        color: $primary-orange;
    }

    p {
        white-space: pre-line;
    }

    span.info {
        color: $dark-grey;
        font-size: 14px;
    }

    div.details {
        display: flex;
        justify-content: space-between;
    }

    div.services {
        width: 50%;
    }

    @media (max-width: $tablet-max) {
        padding: 32px 15px;

        div.details {
            flex-direction: column;
        }
        div.services {
            width: 100%;
        }

        div.members {
            margin: 20px 0 0 0;
        }
    }


    ul.services li {
        margin-top: 15px;
        display: flex;
        padding: 10px;
        border: 1px solid $lighter-grey;
        border-radius: 8px;
        margin-bottom: 15px;
        align-items: center;
        cursor: not-allowed;

        &.uri {
            cursor: pointer;

            &:hover {
                background-color: $light-blue;
                border: 1px solid $light-blue;
            }

        }

        img {
            width: 60px;
            border-radius: 4px;
        }

        span {
            display: inline-block;
            margin-left: 15px;
            padding: 10px;
            border-left: 1px solid $medium-grey;
        }

        svg {
            width: 52px;
            margin: 0 15px 0 auto;
            fill: $lighter-blue;
            color: $lighter-blue;
        }
    }

    .members {
        margin: 0 20px;

        .members-header {
            display: flex;

            &.join-request {
                flex-direction: column;
            }

            a {
                margin-left: 25px;
            }
        }

        span.role {
            color: $dark-grey;
            font-size: 14px;
        }
    }

    .playing-svg {
        margin: auto 0px;

        svg {
            width: 320px;
            height: auto;
        }
    }
}
