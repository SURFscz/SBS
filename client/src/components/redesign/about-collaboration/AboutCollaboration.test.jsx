import React from "react";
import { render } from "@testing-library/react";
import AboutCollaboration from "./AboutCollaboration";

const mockCollaboration = {
    id: "1",
    name: "Test Collaboration",
    description: "This is a test collaboration.",
    services: [],
    members: [],
    collaboration_memberships: []
};

describe("AboutCollaboration", () => {
    it("matches snapshot with default props", () => {
        const { asFragment } = render(
            <AboutCollaboration
            collaboration={mockCollaboration}
            />
        );
        expect(asFragment()).toMatchSnapshot();
    });
});
