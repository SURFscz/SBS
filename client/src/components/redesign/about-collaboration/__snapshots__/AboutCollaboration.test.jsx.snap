// Jest Snapshot v1, https://goo.gl/fbAQLP

exports[`AboutCollaboration matches snapshot with default props 1`] = `
<DocumentFragment>
  <div
    class="collaboration-about-mod"
  >
    <div
      class="about"
    >
      <p
        class="description"
      >
        This is a test collaboration.
      </p>
      <div
        class="services"
      >
        <h4
          class="margin"
        >
          There are no applications connected to your collaboration. 
        </h4>
      </div>
    </div>
    <div
      class="members"
    >
      <div
        class="meta-section"
      >
        <div
          class="header"
        >
          <svg>
            short-name.svg
          </svg>
          <p>
            Short name
          </p>
        </div>
        <div
          class="values"
        >
          <span>
            <strong />
          </span>
        </div>
      </div>
    </div>
  </div>
</DocumentFragment>
`;
